### PR TITLE
fix(api/buffer): fix handling of viewport of non-current buffer

### LINKED ISSUE #27720: `nvim_buf_set_lines` adjustment of cursor and of topline is inconsistent
### Problem

```lua
local lines = vim.api.nvim_buf_get_lines(0, 0, -1, false)
vim.api.nvim_buf_set_lines(0, 0, #lines, true, lines)
```

I believe the above code should do nothing to the buffer, but with the latest nightly, buffer is scrolled downwards resulting an inconsistent user experience.

I was not able to bisect the exact commit causing this problem, but I'm pretty sure this was caused by the series of refactoring https://github.com/neovim/neovim/pull/24889 done by @bfredl .

I found several similar issues, but they referred to the issue with some plugin or lsp.format, whereas my _steps to reproduce_ is reproducible without any setup (nvim -u NONE).

### Steps to reproduce

**No config file required.** (nvim -u NONE)

1. Open a file that has more lines than nvim's window height.
3. `G` to go to the bottom.
4. `H` to move the cursor to the top of the window.
5. Run the following code.

```lua
local lines = vim.api.nvim_buf_get_lines(0, 0, -1, false)
vim.api.nvim_buf_set_lines(0, 0, #lines, true, lines)
```

### Result

The buffer and cursor moves down and lands one line above the bottom `scrolloff`. The exact same behavior as doing `zb` after `step 3.`.

### Expected behavior

Do not scroll the buffer, and have no visual change in above case.

### Neovim version (nvim -v)

NVIM v0.10.0-dev-2504+g3971797be1

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

Linux

### Terminal name/version

Wezterm

### $TERM environment variable

tmux-256color

### Installation

Build from source

### LINKED ISSUE #27723: Regression bug with `nvim_buf_set_lines` on Neovim nightly
### Problem

When using Neovim nightly there is a bug with `nvim_buf_set_lines` when the buffer is not the active one. After calling this function, the buffer is scrolled to top, not preserving the current cursor position. When I switch to the buffer with changed lines, the cursor position is restored and the window is scrolled correctly. It works as expected in Neovim 0.9.5.

Nightly:

https://github.com/neovim/neovim/assets/3128467/b4b0b65a-50cd-413e-8663-53638c721679

0.9.5:

https://github.com/neovim/neovim/assets/3128467/b73476c4-a3d7-4a3b-85db-636823847553

### Steps to reproduce

```lua
local lines = {}
for i = 1, 100 do
  lines[i] = tostring(i)
end

vim.api.nvim_buf_set_lines(OTHER_BUFFER_ID, 0, -1, false, lines)
```

1. Run Neovim nightly
2. Run the provided script for your buffer (use `:luafile` for example)
3. Set your cursor position somewhere in the middle of the file.
4. Create a split and run the provided script for the previously configured buffer (make sure to set the correct buffer id, it can't be 0).

Result: the buffer is scrolled to top.

### Expected behavior

The updated buffer should preserve it's cursor position without scrolling to top.

### Neovim version (nvim -v)

v0.10.0-dev-2501+gdbf6be296

### Vim (not Nvim) behaves the same?

-

### Operating system/version

macOS 14.2.1

### Terminal name/version

iTerm2

### $TERM environment variable

screen-256color

### Installation

repo

### PR DESCRIPTION
**Pull request by bfredl**
A lot of functions in move.c only worked for curwin, alternatively took a `wp` arg but still only work if that happens to be curwin.

Refactor those that are needed for update_topline(wp) to work for any window.

fixes #27723
fixes #27720

### CODE REVIEW

**Review by luukvbaal — COMMENTED** [review]
Some comments referencing curwin need updating.

File: src/nvim/cursor.c [hunk 1]
```diff
@@ -76,16 +76,16 @@ int coladvance_force(colnr_T wcol)
 /// beginning at coladd 0.
 ///
 /// @return  OK if desired column is reached, FAIL if not
-int coladvance(colnr_T wcol)
+int coladvance(win_T *wp, colnr_T wcol)
 {
-  int rc = getvpos(&curwin->w_cursor, wcol);
+  int rc = getvpos(&wp->w_cursor, wcol);
```
[the remark this answers is not in the document]
Comment (bfredl) [reply src/nvim/cursor.c:81]: might need a restructure, coladvance2 can be used to both change text or just update the cursor, and those usecases need to be separated.

File: src/nvim/cursor.c [hunk 2]
```diff
@@ -76,25 +76,26 @@ int coladvance_force(colnr_T wcol)
...
 }
 
-/// @param addspaces  change the text to achieve our goal?
+/// @param addspaces  change the text to achieve our goal? only for wp=curwin!
 /// @param finetune  change char offset for the exact column
 /// @param wcol_arg  column to move to (can be negative)
-static int coladvance2(pos_T *pos, bool addspaces, bool finetune, colnr_T wcol_arg)
+static int coladvance2(win_T *wp, pos_T *pos, bool addspaces, bool finetune, colnr_T wcol_arg)
 {
+  assert(wp == curwin || !finetune);
```
Comment (zeertzjq) [inline src/nvim/cursor.c:98]:
```suggestion
  assert(wp == curwin || !addspaces);
```

File: src/nvim/cursor.c [hunk 3]
```diff
@@ -76,25 +76,26 @@ int coladvance_force(colnr_T wcol)
...
-int coladvance(colnr_T wcol)
+int coladvance(win_T *wp, colnr_T wcol)
 {
-  int rc = getvpos(&curwin->w_cursor, wcol);
+  int rc = getvpos(wp, &wp->w_cursor, wcol);
 
   if (wcol == MAXCOL || rc == FAIL) {
-    curwin->w_valid &= ~VALID_VIRTCOL;
+    wp->w_valid &= ~VALID_VIRTCOL;
   } else if (*get_cursor_pos_ptr() != TAB) {
```
Comment (luukvbaal) [inline src/nvim/cursor.c:85]:
```suggestion
  } else if (*(ml_get_buf(wp->w_buffer, wp->w_cursor.lnum) + wp->w_cursor.col) != TAB) {
```

File: src/nvim/move.c [hunk 4]
```diff
@@ -417,9 +417,9 @@ void update_topline(win_T *wp)
           line_count = wp->w_cursor.lnum - wp->w_botline + 1 + (int)(*so_ptr);
         }
         if (line_count <= wp->w_height_inner + 1) {
-          scroll_cursor_bot(scrolljump_value(), false);
```
Comment (zeertzjq) [inline src/nvim/move.c:420]: The `hasFolding()` above needs to be changed to `hasFoldingWin()` as well.

File: src/nvim/move.c [hunk 5]
```diff
@@ -1211,124 +1194,124 @@ void f_virtcol2col(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
...
-    width1 = curwin->w_width_inner - curwin_col_off();
-    width2 = width1 + curwin_col_off2();
+    width1 = wp->w_width_inner - win_col_off(wp);
+    width2 = width1 + win_col_off2(wp);
   }
 
   // Make sure w_topline is at the first of a sequence of folded lines.
-  hasFolding(curwin->w_topline, &curwin->w_topline, NULL);
-  validate_cursor();            // w_wrow needs to be valid
+  hasFolding(wp->w_topline, &wp->w_topline, NULL);
```
Comment (zeertzjq) [inline src/nvim/move.c:1210]:
```suggestion
  hasFoldingWin(wp, wp->w_topline, &wp->w_topline, NULL);
```

File: src/nvim/move.c [hunk 6]
```diff
@@ -1211,124 +1194,124 @@ void f_virtcol2col(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
...
-        curwin->w_topline--;
-        curwin->w_skipcol = 0;
-        curwin->w_topfill = 0;
+        wp->w_topline--;
+        wp->w_skipcol = 0;
+        wp->w_topfill = 0;
         // A sequence of folded lines only counts for one logical line
         linenr_T first;
-        if (hasFolding(curwin->w_topline, &first, NULL)) {
+        if (hasFolding(wp->w_topline, &first, NULL)) {
```
Comment (zeertzjq) [inline src/nvim/move.c:1238]:
```suggestion
        if (hasFoldingWin(wp, wp->w_topline, &first, NULL)) {
```

File: src/nvim/move.c [hunk 7]
```diff
@@ -1211,124 +1194,124 @@ void f_virtcol2col(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
...
+    validate_cheight(wp);
+    wrow += wp->w_cline_height - 1 -
+            wp->w_virtcol / wp->w_width_inner;
   }
   bool moved = false;
-  while (wrow >= curwin->w_height_inner && curwin->w_cursor.lnum > 1) {
+  while (wrow >= wp->w_height_inner && wp->w_cursor.lnum > 1) {
     linenr_T first;
-    if (hasFolding(curwin->w_cursor.lnum, &first, NULL)) {
+    if (hasFolding(wp->w_cursor.lnum, &first, NULL)) {
```
Comment (zeertzjq) [inline src/nvim/move.c:1288]:
```suggestion
    if (hasFoldingWin(wp, wp->w_cursor.lnum, &first, NULL)) {
```

File: src/nvim/move.c [hunk 8]
```diff
@@ -1211,124 +1194,124 @@ void f_virtcol2col(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
...
+      wrow -= plines_win(wp, wp->w_cursor.lnum--, true);
     }
-    curwin->w_valid &=
+    wp->w_valid &=
       ~(VALID_WROW|VALID_WCOL|VALID_CHEIGHT|VALID_CROW|VALID_VIRTCOL);
     moved = true;
   }
   if (moved) {
     // Move cursor to first line of closed fold.
     foldAdjustCursor();
```
Comment (zeertzjq) [inline src/nvim/move.c:1304]: This also needs de-curwinifying.

File: src/nvim/move.c [hunk 9]
```diff
@@ -1337,153 +1320,153 @@ bool scrolldown(linenr_T line_count, int byfold)
...
-      if (curwin->w_topfill > 0) {
-        curwin->w_topfill--;
+      if (wp->w_topfill > 0) {
+        wp->w_topfill--;
       } else {
-        linenr_T lnum = curwin->w_topline;
+        linenr_T lnum = wp->w_topline;
         if (byfold) {
           // for a closed fold: go to the last line in the fold
           hasFolding(lnum, NULL, &lnum);
```
Comment (zeertzjq) [inline src/nvim/move.c:1362]:
```suggestion
          hasFoldingWin(wp, lnum, NULL, &lnum);
```

File: src/nvim/move.c [hunk 10]
```diff
@@ -1337,153 +1320,153 @@ bool scrolldown(linenr_T line_count, int byfold)
...
   }
 
-  check_topfill(curwin, false);
+  check_topfill(wp, false);
 
-  if (hasAnyFolding(curwin)) {
+  if (hasAnyFolding(wp)) {
     // Make sure w_topline is at the first of a sequence of folded lines.
-    hasFolding(curwin->w_topline, &curwin->w_topline, NULL);
+    hasFolding(wp->w_topline, &wp->w_topline, NULL);
```
Comment (zeertzjq) [inline src/nvim/move.c:1418]:
```suggestion
    hasFoldingWin(wp, wp->w_topline, &wp->w_topline, NULL);
```

File: src/nvim/move.c [hunk 11]
```diff
@@ -1886,101 +1869,101 @@ void set_empty_rows(win_T *wp, int used)
...
-    for (curwin->w_topline = curwin->w_botline;
-         curwin->w_topline > 1;
-         curwin->w_topline = loff.lnum) {
-      loff.lnum = curwin->w_topline;
-      topline_back_winheight(curwin, &loff, false);
+    for (wp->w_topline = wp->w_botline;
+         wp->w_topline > 1;
+         wp->w_topline = loff.lnum) {
+      loff.lnum = wp->w_topline;
+      topline_back_winheight(wp, &loff, false);
```
Comment (zeertzjq) [inline src/nvim/move.c:1894]: The `hasFolding()` in `topline_back_winheight()` should be changed to `hasFoldingWin()`.

File: src/nvim/move.c [hunk 12]
```diff
@@ -1761,54 +1744,54 @@ void scroll_cursor_top(int min_scroll, int always)
...
-  if (curwin->w_cursor.lnum < curwin->w_topline) {
+  int used = wp->w_cline_height;  // includes filler lines above
+  if (wp->w_cursor.lnum < wp->w_topline) {
     scrolled = used;
   }
 
   linenr_T top;  // just above displayed lines
   linenr_T bot;  // just below displayed lines
-  if (hasFolding(curwin->w_cursor.lnum, &top, &bot)) {
+  if (hasFolding(wp->w_cursor.lnum, &top, &bot)) {
```
Comment (zeertzjq) [inline src/nvim/move.c:1756]:
```suggestion
  if (hasFoldingWin(wp, wp->w_cursor.lnum, &top, &bot)) {
```

File: src/nvim/move.c [hunk 13]
```diff
@@ -1761,54 +1744,54 @@ void scroll_cursor_top(int min_scroll, int always)
...
   // "used" already contains the number of filler lines above, don't add it
   // again.
   // Hide filler lines above cursor line by adding them to "extra".
-  int extra = win_get_fill(curwin, curwin->w_cursor.lnum);
+  int extra = win_get_fill(wp, wp->w_cursor.lnum);
 
   // Check if the lines from "top" to "bot" fit in the window.  If they do,
   // set new_topline and advance "top" and "bot" to include more lines.
   while (top > 0) {
     int i = hasFolding(top, &top, NULL)
```
Comment (zeertzjq) [inline src/nvim/move.c:1773]:
```suggestion
    int i = hasFoldingWin(wp, top, &top, NULL)
```

File: src/nvim/move.c [hunk 14]
```diff
@@ -1761,54 +1744,54 @@ void scroll_cursor_top(int min_scroll, int always)
...
     // If scrolling is needed, scroll at least 'sj' lines.
-    if ((new_topline >= curwin->w_topline || scrolled > min_scroll) && extra >= off) {
+    if ((new_topline >= wp->w_topline || scrolled > min_scroll) && extra >= off) {
       break;
     }
 
     used += i;
-    if (extra + i <= off && bot < curbuf->b_ml.ml_line_count) {
+    if (extra + i <= off && bot < wp->w_buffer->b_ml.ml_line_count) {
       if (hasFolding(bot, NULL, &bot)) {
```
Comment (zeertzjq) [inline src/nvim/move.c:1787]:
```suggestion
      if (hasFoldingWin(wp, bot, NULL, &bot)) {
```

File: src/nvim/move.c [hunk 15]
```diff
@@ -1992,145 +1975,145 @@ void scroll_cursor_bot(int min_scroll, bool set_topbot)
   // - scrolled nothing or at least 'sj' lines
   // - at least 'so' lines below the cursor
   // - lines between botline and cursor have been counted
-  if (!hasFolding(curwin->w_cursor.lnum, &loff.lnum, &boff.lnum)) {
+  if (!hasFolding(wp->w_cursor.lnum, &loff.lnum, &boff.lnum)) {
```
Comment (zeertzjq) [inline src/nvim/move.c:1978]:
```suggestion
  if (!hasFoldingWin(wp, wp->w_cursor.lnum, &loff.lnum, &boff.lnum)) {
```

File: src/nvim/move.c [hunk 16]
```diff
@@ -1992,145 +1975,145 @@ void scroll_cursor_bot(int min_scroll, bool set_topbot)
...
-void scroll_cursor_halfway(bool atend, bool prefer_above)
+void scroll_cursor_halfway(win_T *wp, bool atend, bool prefer_above)
 {
-  linenr_T old_topline = curwin->w_topline;
-  lineoff_T loff = { .lnum = curwin->w_cursor.lnum };
-  lineoff_T boff = { .lnum = curwin->w_cursor.lnum };
+  linenr_T old_topline = wp->w_topline;
+  lineoff_T loff = { .lnum = wp->w_cursor.lnum };
+  lineoff_T boff = { .lnum = wp->w_cursor.lnum };
   hasFolding(loff.lnum, &loff.lnum, &boff.lnum);
```
Comment (zeertzjq) [inline src/nvim/move.c:2101]:
```suggestion
  hasFoldingWin(wp, loff.lnum, &loff.lnum, &boff.lnum);
```

File: src/nvim/move.c [hunk 17]
```diff
@@ -2216,70 +2199,70 @@ void scroll_cursor_halfway(bool atend, bool prefer_above)
     }
   }
 
-  if (!hasFolding(topline, &curwin->w_topline, NULL)
-      && (curwin->w_topline != topline || skipcol != 0 || curwin->w_skipcol != 0)) {
-    curwin->w_topline = topline;
+  if (!hasFolding(topline, &wp->w_topline, NULL)
```
Comment (zeertzjq) [inline src/nvim/move.c:2202]:
```suggestion
  if (!hasFoldingWin(wp, topline, &wp->w_topline, NULL)
```

File: src/nvim/move.c [hunk 18]
```diff
@@ -2289,52 +2272,52 @@ void cursor_correct(void)
...
+  linenr_T topline = wp->w_topline;
+  linenr_T botline = wp->w_botline - 1;
   // count filler lines as context
-  int above = curwin->w_topfill;  // screen lines above topline
-  int below = curwin->w_filler_rows;  // screen lines below botline
+  int above = wp->w_topfill;  // screen lines above topline
+  int below = wp->w_filler_rows;  // screen lines below botline
   while ((above < above_wanted || below < below_wanted) && topline < botline) {
     if (below < below_wanted && (below <= above || above >= above_wanted)) {
       if (hasFolding(botline, &botline, NULL)) {
```
Comment (zeertzjq) [inline src/nvim/move.c:2282]:
```suggestion
      if (hasFoldingWin(wp, botline, &botline, NULL)) {
```

File: src/nvim/move.c [hunk 19]
```diff
@@ -2289,52 +2272,52 @@ void cursor_correct(void)
...
         below++;
       } else {
-        below += plines_win(curwin, botline, true);
+        below += plines_win(wp, botline, true);
       }
       botline--;
     }
     if (above < above_wanted && (above < below || below >= below_wanted)) {
       if (hasFolding(topline, NULL, &topline)) {
```
Comment (zeertzjq) [inline src/nvim/move.c:2290]:
```suggestion
      if (hasFoldingWin(wp, topline, NULL, &topline)) {
```

**Review by zeertzjq — COMMENTED** [review]
`hasFoldingWin(wp)` is needed in many places.

**Comment by zeertzjq** [comment]
Or maybe `hasFolding()` can take a `wp` argument, but `hasFoldingWin()` remains unchanged as it is a bit verbose (has two more arguments).

**Comment by bfredl** [comment]
That sounds good, many existing calls can be updated to the shorter form as well.

File: src/nvim/fold.c [hunk 20]
```diff
@@ -1028,9 +1028,9 @@ void foldAdjustVisual(void)
 
 // cursor_foldstart() {{{2
 /// Move the cursor to the first line of a closed fold.
-void foldAdjustCursor(void)
+void foldAdjustCursor(win_T *wp)
 {
-  hasFolding(curwin->w_cursor.lnum, &curwin->w_cursor.lnum, NULL);
+  hasFolding(wp, curwin->w_cursor.lnum, &curwin->w_cursor.lnum, NULL);
```
Comment (zeertzjq) [inline src/nvim/fold.c:1033]:
```suggestion
  hasFolding(wp, wp->w_cursor.lnum, &wp->w_cursor.lnum, NULL);
```

File: src/nvim/ex_docmd.c [hunk 21]
```diff
@@ -5556,23 +5554,19 @@ static void ex_syncbind(exarg_T *eap)
 
   // Set all scrollbind windows to the same topline.
   FOR_ALL_WINDOWS_IN_TAB(wp, curtab) {
-    curwin = wp;
     if (curwin->w_p_scb) {
```
Comment (zeertzjq) [inline src/nvim/ex_docmd.c:5557]:
```suggestion
    if (wp->w_p_scb) {
```

File: src/nvim/cursor.c [hunk 22]
```diff
@@ -76,25 +76,26 @@ int coladvance_force(colnr_T wcol)
...
   }
   return rc;
 }
 
-/// @param addspaces  change the text to achieve our goal?
+/// @param addspaces  change the text to achieve our goal? only for wp=curwin!
 /// @param finetune  change char offset for the exact column
 /// @param wcol_arg  column to move to (can be negative)
-static int coladvance2(pos_T *pos, bool addspaces, bool finetune, colnr_T wcol_arg)
+static int coladvance2(win_T *wp, pos_T *pos, bool addspaces, bool finetune, colnr_T wcol_arg)
```
Comment (zeertzjq) [inline src/nvim/cursor.c:96]: The `mark_mb_adjustpos()` in this function should use `wp->w_buffer` instead of `curbuf`.

File: src/nvim/move.c [hunk 23]
```diff
@@ -2216,70 +2193,70 @@ void scroll_cursor_halfway(bool atend, bool prefer_above)
...
-    int max_off = curwin->w_height_inner / 2;
+    int max_off = wp->w_height_inner / 2;
     if (below_wanted > max_off) {
       below_wanted = max_off;
     }
   }
-  validate_botline(curwin);
-  if (curwin->w_botline == curbuf->b_ml.ml_line_count + 1
+  validate_botline(wp);
+  if (wp->w_botline == curbuf->b_ml.ml_line_count + 1
```
Comment (zeertzjq) [inline src/nvim/move.c:2237]:
```suggestion
  if (wp->w_botline == wp->w_buffer->b_ml.ml_line_count + 1
```

File: src/nvim/move.c [hunk 24]
```diff
@@ -2289,52 +2266,52 @@ void cursor_correct(void)
...
-      curwin->w_valid &=
+    if (cln < topline && wp->w_topline > 1) {
+      wp->w_cursor.lnum = topline;
+      wp->w_valid &=
         ~(VALID_WROW|VALID_WCOL|VALID_CHEIGHT|VALID_CROW);
     }
-    if (cln > botline && curwin->w_botline <= curbuf->b_ml.ml_line_count) {
-      curwin->w_cursor.lnum = botline;
-      curwin->w_valid &=
+    if (cln > botline && wp->w_botline <= curbuf->b_ml.ml_line_count) {
```
Comment (zeertzjq) [inline src/nvim/move.c:2307]:
```suggestion
    if (cln > botline && wp->w_botline <= wp->w_buffer->b_ml.ml_line_count) {
```

**Comment by zeertzjq** [comment]
It seems that a `nvim/types_defs.h` is needed in `nvim/state.h`.

File: src/nvim/cursor.c [hunk 25]
```diff
@@ -104,30 +105,30 @@ static int coladvance2(pos_T *pos, bool addspaces, bool finetune, colnr_T wcol_a
                  || (State & MODE_TERMINAL)
                  || restart_edit != NUL
                  || (VIsual_active && *p_sel != 'o')
-                 || ((get_ve_flags() & VE_ONEMORE) && wcol < MAXCOL);
+                 || ((get_ve_flags(curwin) & VE_ONEMORE) && wcol < MAXCOL);
```
Comment (zeertzjq) [inline src/nvim/cursor.c:108]:
```suggestion
                 || ((get_ve_flags(wp) & VE_ONEMORE) && wcol < MAXCOL);
```

File: src/nvim/cursor.c [hunk 26]
```diff
@@ -159,14 +160,14 @@ static int coladvance2(pos_T *pos, bool addspaces, bool finetune, colnr_T wcol_a
     // is needed to ensure that a virtual position off the end of
     // a line has the correct indexing.  The one_more comparison
     // replaces an explicit add of one_more later on.
-    if (col > wcol || (!virtual_active() && one_more == 0)) {
+    if (col > wcol || (!virtual_active(curwin) && one_more == 0)) {
```
Comment (zeertzjq) [inline src/nvim/cursor.c:163]:
```suggestion
    if (col > wcol || (!virtual_active(wp) && one_more == 0)) {
```

File: src/nvim/cursor.c [hunk 27]
```diff
@@ -159,14 +160,14 @@ static int coladvance2(pos_T *pos, bool addspaces, bool finetune, colnr_T wcol_a
...
       idx -= 1;
       // Don't count the chars from 'showbreak'.
       csize -= head;
       col -= csize;
     }
 
-    if (virtual_active()
+    if (virtual_active(curwin)
```
Comment (zeertzjq) [inline src/nvim/cursor.c:170]:
```suggestion
    if (virtual_active(wp)
```

File: src/nvim/cursor.c [hunk 28]
```diff
@@ -338,19 +339,13 @@ void check_cursor_lnum(win_T *win)
...
-
 /// Make sure win->w_cursor.col is valid. Special handling of insert-mode.
 /// @see mb_check_adjust_col
-void check_cursor_col_win(win_T *win)
+void check_cursor_col(win_T *win)
 {
   colnr_T oldcol = win->w_cursor.col;
   colnr_T oldcoladd = win->w_cursor.col + win->w_cursor.coladd;
-  unsigned cur_ve_flags = get_ve_flags();
+  unsigned cur_ve_flags = get_ve_flags(curwin);
```
Comment (zeertzjq) [inline src/nvim/cursor.c:348]:
```suggestion
  unsigned cur_ve_flags = get_ve_flags(win);
```

File: src/nvim/cursor.c [hunk 29]
```diff
@@ -363,7 +358,7 @@ void check_cursor_col_win(win_T *win)
     if ((State & MODE_INSERT) || restart_edit
         || (VIsual_active && *p_sel != 'o')
         || (cur_ve_flags & VE_ONEMORE)
-        || virtual_active()) {
+        || virtual_active(curwin)) {
```
Comment (zeertzjq) [inline src/nvim/cursor.c:361]:
```suggestion
        || virtual_active(win)) {
```

File: src/nvim/plines.c [hunk 30]
```diff
@@ -540,7 +540,7 @@ void getvcol(win_T *wp, pos_T *pos, colnr_T *start, colnr_T *cursor, colnr_T *en
     if (ci.chr.value == TAB
         && (State & MODE_NORMAL)
         && !wp->w_p_list
-        && !virtual_active()
+        && !virtual_active(curwin)
```
Comment (zeertzjq) [inline src/nvim/plines.c:543]:
```suggestion
        && !virtual_active(wp)
```

File: src/nvim/drawline.c [hunk 31]
```diff
@@ -2342,7 +2342,7 @@ int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, int col_rows, s
               && wlv.line_attr == 0
               && wlv.line_attr_lowprio == 0) {
             // In virtualedit, visual selections may extend beyond end of line
-            if (!(area_highlighting && virtual_active()
+            if (!(area_highlighting && virtual_active(curwin)
```
Comment (zeertzjq) [inline src/nvim/drawline.c:2345]:
```suggestion
            if (!(area_highlighting && virtual_active(wp)
```

File: src/nvim/fold.c [hunk 32]
```diff
@@ -156,9 +156,9 @@ int hasAnyFolding(win_T *win)
 /// lnum of the sequence of folded lines (skipped when NULL).
 ///
 /// @return  true if line "lnum" in the current window is part of a closed fold.
```
Comment (zeertzjq) [inline src/nvim/fold.c:158]:
```suggestion
/// @return  true if line "lnum" in the window "win" is part of a closed fold.
```

Comment (zeertzjq) [inline src/nvim/fold.c:158]:
```suggestion
/// @return  true if line "lnum" in the window "win" is part of a closed fold.
```

Comment (zeertzjq) [inline src/nvim/fold.c:158]:
```suggestion
/// @return  true if line "lnum" in window "win" is part of a closed fold.
```

**Pull request merged**